// ts2esm/dist/src/codemod/convertTSConfig.js
import { select } from '@inquirer/prompts';
import { ModuleKind, ModuleResolutionKind } from 'ts-morph';
import { applyModification } from './applyModification.js';
function isModuleESM(module) {
    switch (module) {
        case ModuleKind.ESNext:
        case ModuleKind.Node16:
        case ModuleKind.NodeNext:
        case ModuleKind.Preserve:
            return true;
        default:
            return false;
    }
}
function isModuleResolutionESM(moduleResolution) {
    switch (moduleResolution) {
        case ModuleResolutionKind.Bundler:
        case ModuleResolutionKind.Node16:
        case ModuleResolutionKind.NodeNext:
            return true;
        default:
            return false;
    }
}
function getAnswer(property, noValue, yesValue) {
    return select({
        choices: [
            {
                name: 'No, I am NOT using a bundler (common for backend apps that rely on Node.js)',
                value: noValue,
            },
            {
                name: 'Yes, I am using a bundler (common for frontend apps that rely on Webpack or Turbopack)',
                value: yesValue,
            },
        ],
        message: `Your "${property}" compiler setting in your TS config needs to be adjusted for ESM.\r\nAre you using a code bundler when compiling?`,
    });
}
export async function convertTSConfig(tsConfigFilePath, project) {
    const isWrongModule = !isModuleESM(project.getCompilerOptions().module);
    const isWrongModuleResolution = !isModuleResolutionESM(project.getCompilerOptions().moduleResolution);
    if (isWrongModule) {
        const module = await getAnswer('module', 'nodenext', 'esnext');
        await applyModification(tsConfigFilePath, '/compilerOptions/module', module);
    }
    if (isWrongModuleResolution) {
        const moduleResolution = await getAnswer('moduleResolution', 'nodenext', 'bundler');
        await applyModification(tsConfigFilePath, '/compilerOptions/moduleResolution', moduleResolution);
    }
}
