// ts2esm/dist/src/util/PathUtil.js
import path from 'node:path';
import fs from 'node:fs';
export function isMatchingPath(pattern, path) {
    const normalized = removeWildCards(pattern);
    if (path.startsWith(normalized)) {
        return pattern.length;
    }
    return 0;
}
export function removeWildCards(pattern) {
    return pattern.replace('/*', '').replace('*', '');
}
export function removePathAlias(alias, path) {
    return path.replace(removeWildCards(alias), '').replace('/', '');
}
export function findBestMatch(aliasMap, path) {
    let bestRank = 0;
    let bestMatch = '';
    for (const key of Object.keys(aliasMap)) {
        const rank = isMatchingPath(key, path);
        if (rank > bestRank) {
            bestRank = rank;
            bestMatch = key;
        }
    }
    return bestMatch;
}
export function getNormalizedPath(projectDirectory, info, paths) {
    const pathAliasResolution = `${paths[info.pathAlias]}`;
    const normalizedResolution = removeWildCards(pathAliasResolution).replaceAll(info.quoteSymbol, '');
    const normalizedFilePath = removePathAlias(info.pathAlias, info.normalized);
    return path.join(projectDirectory, normalizedResolution, normalizedFilePath);
}
export function hasRelativePath(path) {
    if (path === '.' || path === '..') {
        return true;
    }
    return path.startsWith('./') || path.startsWith('../');
}
export function isNodeModuleRoot(directory) {
    const isInNodeModules = directory.includes(`node_modules${path.sep}`);
    const packageJsonPath = path.join(directory, 'package.json');
    const packageJsonExists = fs.existsSync(packageJsonPath);
    return isInNodeModules && packageJsonExists;
}
